tools/autograd/templates/variable_factories.h

```diff
@@ -22,80 +22,170 @@ using at::DimnameList;
+// to the equivalent Tensor successfully. Consider the following example:
+//
+// `torch::tensor({{1}, {2}})`
+//
+// Here is the code path that it goes through:
+//
+// `at::Tensor tensor(detail::TensorDataContainer<1> init_list_tensor)`
+//
+// which calls:
+//
+// `TensorDataContainer<1>({{1}, {2}})`
+//
+// which matches to the `TensorDataContainer<1>(std::initializer_list<TensorDataContainer<2>>)` constructor,
+// and in an attempt to convert `{1}` and `{2}` to `TensorDataContainer<2>`, it calls the following:
+//
+// `TensorDataContainer<2>({1})`  (same call path happens for `{2}`, and we'll just focus on `{1}` here)
+//
+// At this point, theoretically there are two plausible ways for `{1}` to be matched to one of the
+// constructors of `TensorDataContainer<2>`:
+//
+// 1. It can be a list-initialization of a scalar value, thus matching `TensorDataContainer<2>(int value)`.
+// 2. It can be converted to `std::initializer_list<TensorDataContainer<3>>`, thus matching
+//    `TensorDataContainer<2>(std::initializer_list<TensorDataContainer<3>>)`.
+//
+// How does the compiler decide which one to choose? According to `https://en.cppreference.com/w/cpp/language/list_initialization`,
+// braced-init-list always prefers the constructor that takes `std::initializer_list`. Hence we happily
+// move forward with constructor #2, and it calls the following:
+//
+// `TensorDataContainer<3>(1)`
+//
+// Now it matches `TensorDataContainer<3>(int value)`, which stores `1` as a scalar value. All is good.
+//
+// Note that `torch::tensor({{1}, {2}})` can also match another existing function overload:
+// `torch::tensor(at::ArrayRef<int> values)`, because `{1}` and `{2}` can be treated as
+// a list-initialization of an `int` value. However, this will produce a Tensor with sizes `{2}`,
+// but we actually want a Tensor with sizes `{2, 1}`. In order to avoid matching this function overload,
+// we move the ability to convert `at::ArrayRef<T>` (and similarly `std::vector<T>`) into `TensorDataContainer<1>`,
+// and since for braced-init-list the `TensorDataContainer<1>(std::initializer_list<TensorDataContainer<2>>)`
+// constructor is always preferred over all other constructors, it will take the `std::initializer_list` path
+// and all is good again.
+//
+// P.S. You might ask: why do we need to templatize `TensorDataContainer`, and why can't we just provide a
+// `TensorDataContainer(std::initializer_list<TensorDataContainer>>)` constructor to reach the innermost
+// dimension of an arbitrarily nested braced-init-list? The answer is that for a braced-init-list
+// like `{{1}, {2}}`, when the compiler is handling the inner braced-init-list `{1}` and `{2}`, it seems
+// to always prefer the scalar constructor `TensorDataContainer(int)` instead of the `std::initializer_list`
+// constructor, thus producing a tensor with the wrong sizes. If we templatize `TensorDataContainer` over
+// the # of tensor dimensions, such problem doesn't happen.
+template <size_t D>
```

Comment:
What is D? The depth? Could we name it something more descriptive?

Reply:
I refactored the PR and got rid of the need for this template variable :D